File: Solution-Java/Design_211.java
class WordDictionary {
    
    class TrieNode {
        private TrieNode[] next = new TrieNode[26];
        private boolean isEnd = false;
    }
    
    private TrieNode root;

    /** Initialize your data structure here. */
    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            if (node.next[c - 'a'] == null) {
                node.next[c - 'a'] = new TrieNode();
            }
            node = node.next[c - 'a'];
        }
        node.isEnd = true;
    }
    
    public boolean search(String word) {
        return search(word, 0, root);
    }
    
    private boolean search(String word, int start, TrieNode node) {
        for (int i = start; i < word.length(); i++) {
            char c = word.charAt(i);
            if (c == '.') {
                for (int j = 0; j < 26; j++) {
                    if (node.next[j] != null && search(word, i + 1, node.next[j])) {
                        return true;
                    }
                }
                return false;
            } else if (node.next[c - 'a'] == null) {
                return false;
            } else {
                node = node.next[c - 'a'];
            }
        }
        return node.isEnd;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: Solution-Python/Design_211.py
class TrieNode:
    def __init__(self):
        self.next = dict()
        self.is_end = False

class WordDictionary:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for c in word:
            if not node.next.get(c):
                node.next[c] = TrieNode()
            node = node.next[c]
        node.is_end = True

    def search(self, word: str) -> bool:
        return self.search_helper(word, 0, self.root)
        
    def search_helper(self, word: str, start: int, node: TrieNode) -> bool:
        for i in range(start, len(word)):
            c = word[i]
            if c == '.':
                for j in string.ascii_lowercase:
                    if node.next.get(j) and self.search_helper(word, i + 1, node.next[j]):
                        return True
                return False
            elif not node.next.get(c):
                return False
            else:
                node = node.next[c]
        return node.is_end


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)